bundle/terranova/plan.go

```diff
@@ -48,6 +54,153 @@ func (d *DeploymentUnit) plan(ctx context.Context, client *databricks.WorkspaceC
 	return calcDiff(d.Adapter, savedState, config)
 }
 
+func (d *DeploymentUnit) refreshRemoteState(ctx context.Context, id string) error {
+	if d.Fresh {
+		return nil
+	}
+
+	// TODO retry failures
+	remoteState, err := d.Adapter.DoRefresh(ctx, id)
+	if err != nil {
+		return fmt.Errorf("failed to refresh remote state id=%s: %w", id, err)
+	}
+	err = d.SetRemoteState(remoteState)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type ErrorDelayed struct{}
+
+func (e ErrorDelayed) Error() string { return "must be resolved after apply" }
+
+var ErrDelayed ErrorDelayed
+
+func (d *DeploymentUnit) ResolveReferenceLocalOrRemote(ctx context.Context, db *tnstate.TerranovaState, reference string, actionType deployplan.ActionType, config any) (any, error) {
+	path, ok := dynvar.PureReferenceToPath(reference)
+	if !ok || len(path) <= 3 || path[0].Key() != "resources" || path[1].Key() != d.Group || path[2].Key() != d.Key {
+		return nil, fmt.Errorf("internal error: expected reference to resources.%s.%s, got %q", d.Group, d.Key, reference)
+	}
+
+	fieldPath := path[3:]
+
+	if fieldPath.String() == "id" {
+		if actionType.KeepsID() {
+			entry, hasEntry := db.GetResourceEntry(d.Group, d.Key)
+			idValue := entry.ID
+			if !hasEntry || idValue == "" {
+				return nil, errors.New("internal error: no db entry")
+			}
+			return idValue, nil
+		}
+		// id may change after deployment, this needs to be done later
+		return nil, ErrDelayed
+	}
+
+	// Note: using config instead of resource.Config() means we have access too full schema besides SDK, e.g. permissions or grants.
+	configValidErr := structaccess.Validate(reflect.TypeOf(config), fieldPath)
+	remoteValidErr := structaccess.Validate(d.Adapter.RemoteType(), fieldPath)
+
+	if configValidErr != nil && remoteValidErr != nil {
+		return nil, fmt.Errorf("schema mismatch: %w; %w", configValidErr, remoteValidErr)
+	}
+
+	if configValidErr == nil && remoteValidErr != nil {
+		// The field is only present in local schema; it must be resolved here.
+		value, err := structaccess.Get(config, fieldPath)
+		if err != nil {
+			return nil, fmt.Errorf("field not set: %w", err)
+		}
+
+		return value, nil
+	}
+
+	if configValidErr != nil && remoteValidErr == nil {
+		// The field is only present in remote state schema.
+		// If resource is unchanged in this plan, we can proceed with resolution.
+		// If resource is going to change, we need to postpone this until deploy.
```

Comment:
Does the action type refer to the one for this deployment unit, or the one that it is referring to?